### ISSUE TITLE
Potential issue when checking the return code of a publish

### ISSUE THREAD

**Issue by rulez85** [eclipse-paho/paho.mqtt.c#240]
Hi all,

I found a potential issue when checking the return code of a publish.
This is my configuration:
- OS: Win32
- Compiler: MSVC
- Paho version: 1.1.0
- Paho library: paho-mqtt3cs (synchronous version of the c client library with SSL)
- OpenSSL version: 1.0.1-u

According to the "MQTTClient_publish" example, I was under the impression that the return value of the "MQTTClient_waitForCompletion" method could be used to determine whether a message has been successfully published to the broker (usign QoS > 0).
Am I right?

If this is the case, then I think I hit a bug when I try to perform a publish after turning off the connection (e.g., unplugging the ethernet cable).

I also seen that there is another API "MQTTClient_isConnected" but its documentation does not fully explain when and how it works.

Here is a simple test:

```cpp
#include "stdio.h"
#include "stdlib.h"
#include "string.h"
#include "MQTTClient.h"

#define TEST_SSL

#ifdef TEST_SSL
#define ADDRESS     "ssl://test.mosquitto.org:8883"
#else
#define ADDRESS     "tcp://test.mosquitto.org" // using default no-ssl port
#endif

#define CLIENTID    "MyClientID"
#define TOPIC       "MyTOPIC"
#define PAYLOAD     "Hello World!"
#define QOS         1
#define TIMEOUT     10000L

int main(int argc, char* argv[])
{
    MQTTClient client;
    MQTTClient_connectOptions conn_opts = MQTTClient_connectOptions_initializer;
    MQTTClient_message pubmsg = MQTTClient_message_initializer;
    MQTTClient_deliveryToken token;
    int rc;

    MQTTClient_create(&client, ADDRESS, CLIENTID,
        MQTTCLIENT_PERSISTENCE_NONE, NULL);
    conn_opts.keepAliveInterval = 20;
    conn_opts.cleansession = 1;

#ifdef TEST_SSL
    MQTTClient_SSLOptions sslOptions = MQTTClient_SSLOptions_initializer;
    sslOptions.enableServerCertAuth = 1;
    sslOptions.trustStore = "C:\\paho\\build-vc140-win32\\src\\samples\\Debug\\mosquitto.org.crt";
    conn_opts.ssl = &sslOptions;
#endif

    if ((rc = MQTTClient_connect(client, &conn_opts)) != MQTTCLIENT_SUCCESS)
    {
        printf("Failed to connect, return code %d\n", rc);
        exit(EXIT_FAILURE);
    }
    pubmsg.payload = PAYLOAD;
    pubmsg.payloadlen = strlen(PAYLOAD);
    pubmsg.qos = QOS;
    pubmsg.retained = 0;

    // ///////////////////////////
    // step 1: connection on
    MQTTClient_publishMessage(client, TOPIC, &pubmsg, &token);
    printf("Waiting for up to %d seconds for publication of %s\n"
        "on topic %s for client with ClientID: %s\n",
        (int)(TIMEOUT / 1000), PAYLOAD, TOPIC, CLIENTID);
    rc = MQTTClient_waitForCompletion(client, token, TIMEOUT);

    if (rc == MQTTCLIENT_SUCCESS)
        printf("Publish #1 success!\n");
    else
        printf("Publish #1 failed!\n");

    printf("MQTTClient_isConnected = %d\n", MQTTClient_isConnected(client));

    // ///////////////////////////
    // step 2:  connection off
    printf("Press a key after turn off the connection");
    getchar();

    MQTTClient_publishMessage(client, TOPIC, &pubmsg, &token);
    printf("Waiting for up to %d seconds for publication of %s\n"
        "on topic %s for client with ClientID: %s\n",
        (int)(TIMEOUT / 1000), PAYLOAD, TOPIC, CLIENTID);
    rc = MQTTClient_waitForCompletion(client, token, TIMEOUT);

    if (rc == MQTTCLIENT_SUCCESS)
        printf("Publish #2 success!\n");
    else
        printf("Publish #2 failed!\n");

    printf("MQTTClient_isConnected = %d\n", MQTTClient_isConnected(client));

    MQTTClient_disconnect(client, 10000);
    MQTTClient_destroy(&client);
    return rc;
}

```

Here is the application output:
```
Waiting for up to 10 seconds for publication of Hello World!
on topic MyTOPIC for client with ClientID: MyClientID
Publish #1 success!
MQTTClient_isConnected = 1
Press a key after turn off the connection
Waiting for up to 10 seconds for publication of Hello World!
on topic MyTOPIC for client with ClientID: MyClientID
Publish #2 success!
MQTTClient_isConnected = 0
```

As you can see, the return value "rc" of the "MQTTClient_waitForCompletion" method evaluates incorrectly to "MQTTCLIENT_SUCCESS", which cannot be possible since the connection has
been interrupted by removing the ethernet cable, while "MQTTClient_isConnected" evaluates to "false".

If I comment the "#define TEST_SSL" line in the program in order not to use the SSL version, I get the opposite application output:
```
Waiting for up to 10 seconds for publication of Hello World!
on topic MyTOPIC for client with ClientID: MyClientID
Publish #1 success!
MQTTClient_isConnected = 1
Press a key after turn off the connection
Waiting for up to 10 seconds for publication of Hello World!
on topic MyTOPIC for client with ClientID: MyClientID
Publish #2 failed!
MQTTClient_isConnected = 1
```
The return value "rc" of the "MQTTClient_waitForCompletion" method evaluates correctly to "-1", while "MQTTClient_isConnected" evaluates incorrectly to "true", which does not appear
to be right since the connection has been interrupted.

What is the proper way to check whether a publish has been successful (with QOS 1)?

Thanks in advance,

Nicola

**Comment by rulez85** [comment]
Here http://www.eclipse.org/paho/files/mqttdoc/Cclient/async.html  is written:
``
In synchronous mode, the client application runs on a single thread. Messages are published using the MQTTClient_publish() and MQTTClient_publishMessage() functions. To determine that a QoS1 or QoS2 (see Quality of service) message has been successfully delivered, the application must call the MQTTClient_waitForCompletion() function.
``

Therefore I think this is a significant bug (even if the documentation does not fully indicates how to check the success of the publish operation) since published messages with QoS 1 has been lost.

**Comment by rulez85** [comment]
Seems the bug appears only if the second publish is performed after the connection keepAliveInterval.
If publish is performed before it (20 seconds in my example), "MQTTClient_waitForCompletion" method evaluates correctly to "-1", otherwise it evaluates incorrectly to "MQTTCLIENT_SUCCESS".

**Comment by CheyenneForbes** [comment]
[ http://www.eclipse.org/paho/files/mqttdoc/Cclient/async.html](http://www.eclipse.org/paho/files/mqttdoc/Cclient/async.html) is giving a 404 error

**Comment by rulez85** [comment]
Seems the paho c client official doxygen page is no more reachable as was few weeks ago... Do not known why, may be a problem on the web site or simply the doxygen has been removed

**Comment by icraggs** [comment]
I updated the online documentation pages, and in the process the links were changed.  They work if followed from the client page: http://www.eclipse.org/paho/clients/c/

The new links are:

http://www.eclipse.org/paho/files/mqttdoc/MQTTClient/html/index.html
http://www.eclipse.org/paho/files/mqttdoc/MQTTAsync/html/index.html

**Comment by rulez85** [comment]
Ok, so the link in the second comment now is http://www.eclipse.org/paho/files/mqttdoc/MQTTClient/html/async.html

What about the bug?

**Comment by icraggs** [comment]
I've added a fix to the develop branch, if you would like to try it out.

**Comment by rulez85** [comment]
Thank you, I tried it and now seems to be ok!

Edit: 
I found a problem with the last develop version (commit eabca5bf2ab4e456f3f1b1d82371d3b3064a331f of April 12th). 
It works correctly within mosquitto broker (test.mosquitto.org) but it doesn't works within Azure IoT Hub broker. In particular, the issue seems to be into the connection phase (`MQTTClient_connect` returns `-1`). Rolling back to the release 1.1.0 the library returns to works correctly.
